transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -842,10 +843,8 @@ public ChannelFuture writeAndFlush(Object msg) {
     }
 
     private static void notifyOutboundHandlerException(Throwable cause, ChannelPromise promise) {
-        if (!promise.tryFailure(cause) && !(promise instanceof VoidChannelPromise)) {
-            if (logger.isWarnEnabled()) {
-                logger.warn("Failed to fail the promise because it's done already: {}", promise, cause);
-            }
+        if (!(promise instanceof VoidChannelPromise)) {
```

Comment:
humm ... fine for this PR ... the usage of `void` promise is infectious and is often not taken into account practically. I think we should re-visit this concept and related APIs in the next release.